Linear algebra operations, Dimensionality reduction and some other minor changes

MustafaKarabulut opened the pull request:
Hi,

For a while, I've been doing a research about dimensionality reduction techniques including PCA, KernelPCA, LDA and Random Projections. So far, I am able to implement and validate the PCA algorithm and currently working on KernelPCA. The latter is also about to be finished but there are something preventing me to validate its results. Nonetheless, I thought it would be harder to review and discuss so many things at the same time and thus it seemed more reasonable to submit them part-by-part.

Here is the first part:
- PCA implementation and its test
- KernelPCA implementation (test not included yet)
- Linear algebra namespace including EigenvalueDecomposition. This was mainly extracted by the PHP port of JAMA which is licenced under PHP v3.0. I don't see any reason that can prevent us using, changing and/or distributing it. So far, I only used EigenvalueDecomposition but in the future I plan to use other decomposition techniques part of the same library (Cholesky, LU, QR and SingularValue).
- I added some static and instance methods to Matrix class whenever I felt a need about them.
- I also needed to add Covariance class that was required in the PCA implementation.

At the moment
- PCA is working
- KernelPCA is somewhat working but there is a consistency problem with fit and transform methods. Moreover, I need an example whose result is known beforehand to validate its results.
- I will pass onto LDA and Random Projections as soon as these are completely done.

Please share your comments and let me know what you think about these.
I'll be updating this PR as soon as I get some progress.

itcraftsmanpl commented on src/Phpml/Math/LinAlg/EigenValueDecomposition.php:
```diff
@@ -0,0 +1,894 @@
+ *  @author Mustafa Karabulut
+ */
+
+namespace Phpml\Math\LinAlg;
```
I think we can change it to `Phpml\Math\LinearAlgebra`

MustafaKarabulut replied:
OK.

itcraftsmanpl commented:
Hi @MustafaKarabulut nice PR. You need to change filename `EigenValueDecomposition.php` because is not compatible with class name in this file `EigenvalueDecomposition`.

MustafaKarabulut commented:
Hi @itcraftsmanpl, I just realized the class name and file name do not match. I think this is the very reason about the failure in Travis CI. I'll fix it.

itcraftsmanpl commented on src/Phpml/Math/Matrix.php:
```diff
@@ -283,4 +356,82 @@ public function isSingular() : bool
+     *
+     * @return array
+     */
+    public static function map(array $array, $callable)
```
we need tests for every new public function in this class

MustafaKarabulut replied:
I was afraid that you would eventually ask for it. But I have to admit you are absolutely right!

itcraftsmanpl commented:
Pleas add at least new line in readme.md (with new features). I now that we need full documentation, but time is priceless and somebody else can do it.

MustafaKarabulut commented:
After I can ensure things in this part work as expected, I will deal with the things you ask for

itcraftsmanpl commented on tests/Phpml/DimensionReduction/KernelPCATest.php:
```diff
@@ -0,0 +1,53 @@
+
+        $kpca = new KernelPCA(KernelPCA::KERNEL_RBF, null, 1, 15);
+        $reducedData = $kpca->fit($data);
+        $transformed2 = (new Matrix($reducedData))->transpose()->toArray()[0];
```
maybe reduced data should be already transponsed? (we want to avoid calling tranponse always after reducing data)

MustafaKarabulut replied:
It is already like that. However, for the test case, I had given the transformation matrix to be compared in one-dimensional array so it somewhat confused the comparison. I fixed it and now transpose is not needed in the test function anymore.

MustafaKarabulut commented:
Latest update:
- Completed PCA and KernelPCA tests
- Updated Matrix test file by adding new test functions for newly added methods in this PR
- Removed some unused methods from Matrix.php that I added in this PR
- Changed some Matrix.php methods (getColumnValue and transpose) in a way that they use php built-in functions instead of loops. This increased the performance substantially as to my performance tests. For instance the current implementation of transpose() takes about 4,1 sec to transposing a 2000x2000 matrix for 10 times while the new version does this job under 2 secs (1,89 sec). About getColumnValues(), its performance is also increased but not so obviously that I observed in transpose() method (Current version: 0,025 sec to get 100 columns of this matrix while the new one takes only 0,009 sec).

MustafaKarabulut commented:
Lastly, I have also updated Readme file to include the recently added algorithms.

MustafaKarabulut commented:
Apperantly Travis CI failed due to something unrelated to the changes in this PR. 

>1) tests\Phpml\NeuralNetwork\Training\BackpropagationTest::testBackpropagationForXORLearning

>  Failed asserting that 0.44910413263687554 matches expected 0.

>  /home/travis/build/php-ai/php-ml/tests/Phpml/NeuralNetwork/Training/BackpropagationTest.php:25

The build passed for PHP 7.0 but it did not for PHP 7.1. In my opinion, the failure may be something related to the random nature of the BP algorithm and may succeed if run one more time.

@itcraftsmanpl do you have an idea about what the problem is?

MustafaKarabulut commented:
It succeeded now. As I guessed, it is probably something about random initialization of some algorithm.

dmonllao commented:
Hi Mustafa, I've experienced this random failure multiple times, I agree it shouldn't be related to this issue.

itcraftsmanpl commented:
@MustafaKarabulut great work :+1: :+1: we need more test but for now is enough.
This test irritates me from the beginning but I can not find a solution.
In performance topic: i think we create separate benchmarks script (or repo) and then everybody can experiment with code and improve performance. But i think now we need more features to release version 1.0 and then we can focus on other aspects.

MustafaKarabulut commented:
@itcraftsmanpl First of all, thanks for accepting the PR.

Also I need to say, my thoughts are similar to yours about developing first and documenting/testing later approach. Sometime in the near future, we'll have completed the most essential of the algorithms in each category (classification, clustering, dimensionality reduction, feature selection and so forth) and we'll be having some time to improve other aspects of the library such as its documentation, sample applications and a nice website.
